Iteration::close

franzpoeschel opened the pull request:
While the openPMD API currently has a data layout (file-based layout) that creates new files for each iteration of data, no explicit method to close such a file currently exists. This PR adds a call `Iteration::close`.

TODO:
- [x] Implement for backends other than ADIOS2
- [x] `Iteration::close` is deferred, like essentially all other calls in the API so far, the file will only be closed upon `flush`ing. 1. I need to document that fact better, 2. should we keep it that way? Maybe something like `close(bool flush = true)` could be helpful?

franzpoeschel committed:
Add IOTask for closing file

franzpoeschel committed:
Add API calls for Iteration::close

franzpoeschel committed:
Add closeFile to ADIOS2 backend

franzpoeschel committed:
remove skipFlush

and other wrongly commited lines

franzpoeschel committed:
Add test

franzpoeschel committed:
add back skipFlush

franzpoeschel committed:
Invalidate file in ADIOS2 after closing

ax3l commented:
Thanks a lot for the PR! I think going for now with `close(bool flush = true)` would be good! Maybe we get #484 at some point for more global control...

ax3l commented on include/openPMD/IO/IOTask.hpp:
```diff
@@ -127,6 +128,22 @@ struct OPENPMDAPI_EXPORT Parameter< Operation::OPEN_FILE > : public AbstractPara
+struct OPENPMDAPI_EXPORT Parameter< Operation::CLOSE_FILE > : public AbstractParameter
+{
+    Parameter() = default;
+    Parameter( Parameter const & p ) : AbstractParameter(), name( p.name ){};
```
```suggestion
    Parameter( Parameter const & p ) : AbstractParameter(), name( p.name ){}
```

ax3l committed:
Remove superfluous ; after function

franzpoeschel committed:
Make Iteration::close synchronous by default

Optional parameter `bool flush = true`

ax3l commented on src/Iteration.cpp:
```diff
@@ -97,6 +97,32 @@ Iteration::setTimeUnitSI(double newTimeUnitSI)
+Iteration &
+Iteration::close( bool flush )
+{
+    setAttribute( "closed", true );
```
do we need an attribute, e.g. for streaming reasons, or would a private member be sufficient?

franzpoeschel replied:
> e.g. for streaming reasons

That's exactly the reason, yes. I use this in the data sink to know which iterations I need not parse any more.

ax3l replied:
Sounds good then :)

franzpoeschel committed:
Add closeFile operation to JSON backend

franzpoeschel committed:
Remove name parameter from CLOSE_FILE task

franzpoeschel committed:
Fix name shadowing

franzpoeschel committed:
[wip] close_file for ADIOS1

somehow still fails when creating the second iteration

franzpoeschel committed:
Temporary workaround for ADIOS1

franzpoeschel commented:
Does ADIOS1 have any issues with having multiple instances (i.e. a writer and a reader) interfere with one another? See the comment in the [latest commit](https://github.com/openPMD/openPMD-api/pull/746/commits/29e5251b34ca3a4897a5b66b862edff8336d4638).

franzpoeschel committed:
Implement file closing in HDF5

franzpoeschel commented:
HDF5 seems to be working, ADIOS1 is still being a bit moody atm..

ax3l commented:
Could be a double free.

ax3l commented on src/IO/ADIOS/CommonADIOS1IOHandler.cpp:
```diff
@@ -540,8 +540,68 @@ CommonADIOS1IOHandlerImpl::openFile(Writable* writable,
+        close( handle_read->second );
+        m_openReadFileHandles.erase( handle_read );
+    }
+    m_filePaths.erase( myFile );
+    m_existsOnDisk.erase( myFile->second );
```
Looks more logical/safe to me:
```suggestion
    m_existsOnDisk.erase( myFile->second );
    m_filePaths.erase( myFile );
```

ax3l committed:
change erase order

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -91,8 +91,80 @@ TEST_CASE( "multi_series_test", "[serial]" )
+    /*
+     * This block will run fine if commented in
+     * But the following block will fail with a segfault
+     * upon adios_select_method ????
```
that's because you cannot initialize ADIOS1 twice in the same process.
ADIOS1 uses some nasty static states.

So: no two series that are active at the same time in a process and use each an ADIOS1 backend.
See: https://openpmd-api.readthedocs.io/en/latest/backends/adios1.html#limitations

franzpoeschel replied:
That explains things, was already suspecting something in this direction. So, the basic "close file" functionality should be working anyway for ADIOS1, but we should probably just run a simplified test, such as checking whether the file is on disk?
Alternatively, running an MPI test with two processes could circumvent that, but I'm feeling that would be overkill...

ax3l replied:
Good idea, for ADIOS1 we can use our helper methods to check we found the files on disk from a single rank instead of reading them back early. Let's keep the logic for the other backends as is, it looks good to me.

franzpoeschel committed:
run a simplified test for ADIOS1

franzpoeschel committed:
Split closed into closed and closedByWriter

This enables closing iterations in read-mode, too.
closedByWriter will be mainly useful in streaming mode, which this
commit is preparing

franzpoeschel committed:
Fix ADIOS1 test

franzpoeschel commented:
Writing attributes to indicate files having been closed has been problematic for read mode. I used a boolean flag on `Iteration` now to circumvent that.
I left the attribute in, since it will be necessary for streaming mode later on anyway.
Also fixed the ADIOS1 test.

franzpoeschel committed:
cleanup

ax3l commented on include/openPMD/Iteration.hpp:
```diff
@@ -97,7 +111,21 @@ class Iteration : public Attributable
+     * Indicates that an iteration has been logically closed.
+     * Will be physically closed upon next flush.
+     */
+    std::shared_ptr< bool > m_closed = std::make_shared< bool >( false );
```
Can you quickly remind why this cannot be a member? Is some lifetime of the series/iteration preventing this?

franzpoeschel replied:
`class Iteration` is designed as a handle, all relevant information should be shared between copies.
```cpp
auto iteration0 = series.iterations[0];
iteration0.close();
```
If the flag were not behind a `shared_ptr`, we would then get:
```cpp
iteration0.m_closed == true
// but
series.iterations[0].m_closed == false
```
(This touches a broader point that we should keep in mind, if we renovate the frontend some day: While most classes in the openPMD API are handles, this is implemented rather in an ad-hoc way currently. I would prefer a design similar to ADIOS2, i.e. have some none-copyable resource like `internal::Iteration` and a user-facing handle `external::Iteration` that acts like a `shared_ptr` to `internal::Iteration`)

ax3l commented on include/openPMD/Iteration.hpp:
```diff
@@ -97,7 +111,21 @@ class Iteration : public Attributable
+     * to avoid reading undefined data if a backend implements optimizations
+     * based on this information.
+     */
+    std::shared_ptr< bool > skipFlush = std::make_shared< bool >( false );
```
why do we need both `m_closed` and `[m_]skipFlush`?

franzpoeschel replied:
Documentation here is outdated, will fix – also maybe the naming should be redone, something like `logicallyClosed` and `physicallyClosed`. `logicallyClosed` means that the iteration has been closed, but the corresponding `flush` has not run yet.

ax3l replied:
then I would just call these states `m_closed` and `m_flushed` or so. (we have no control when something really is done transferring, that is up to the backend we talk to.)

franzpoeschel replied:
Hm, I think `m_flushed` does not sound very clear about what it's doing?
I mean, while you can certainly argue that `m_physicallyClosed` is not *technically* correct, it is "physically closed" from the frontend's perspective. Anything from now on is the backend's problem which the frontend needs not and cannot interfere with any longer. I'd say that's close enough? :D 

Alternatively, what about `m_closedInFrontend` and `m_closedInBackend`?

franzpoeschel replied:
Changed this thing into an enumeration that can take the values `Open`, `ClosedInFrontend` and `ClosedInBackend` now.

ax3l commented on src/Series.cpp:
```diff
@@ -564,6 +564,10 @@ Series::flushFileBased()
         {
+            if ( *i.second.skipFlush )
+            {
+                continue;
```
shouldn't we throw on this in case a users continues to issue API calls on an (accidentally) closed iteration?

franzpoeschel replied:
The flushing procedures haven't been written with the possibility in mind that an iteration might not be present any longer. Flushing an iteration will e.g. always open the corresponding file, irregardless of whether it is still there or not.

ax3l replied:
Probably the suggested renaming above will already help a bit. Anyway, this looks like something that will be very hard to debug.

Maybe I missed this: When would a closed and flushed iteration be flushed again? When we change global parameters of an overall series such as extension at a late point, maybe? Sounds legit to ignore that and we should maybe write some warnings to stderr when this happens.

franzpoeschel replied:
>When would a closed and flushed iteration be flushed again?

The idea behind continuing with the next iteration in that case is exactly to *not* flush it again. I've added sanity checks now to check whether the iteration hasnt been wrongly accessed and we can safely skip it.

franzpoeschel committed:
Throw if flushing a closed iteration that has been accessed

franzpoeschel committed:
Add verifyClosed

franzpoeschel committed:
Close files in read mode too

franzpoeschel committed:
Test throwing upon accessing an iteration post closing

franzpoeschel committed:
Also do sanity-checks in group-based mode

This is currently not necessary, since closing an iteration in
group-based mode is a no-op. It will become relevant in streaming-based
mode, where closing an iteration means discarding its corresponding data
packet.

franzpoeschel committed:
Rename m_closed* members of Iteration and update docu

franzpoeschel committed:
Unify m_closed(Backend|Frontend) into common enum

franzpoeschel committed:
Flush only the closed iteration upon Iteration::close

franzpoeschel committed:
Undo unnecessary whitespace changes

franzpoeschel committed:
Two small fixes in ADIOS1 IO Handler

1) Remove debugging output
2) Delete buffered attribute writes after performing them

franzpoeschel committed:
Prevent throwing in destructor in close_iteration_test

franzpoeschel committed:
Do not use boolean attributes

ax3l commented:
@franzpoeschel please rebase against latest `dev` (or merge it in) for latest CI updates.

ax3l commented:
Ah wait, I actually see a memory leak is caught. Is that from ADIOS2? In case it's not wrong usage on our side that causes this we can report it upstream and add a temporary ignore in `.github/ci/sanitizer/clang/Leak.supp`

franzpoeschel commented:
I intentionally leak memory in the serial IO tests to avoid throwing in a destructor [1](https://github.com/openPMD/openPMD-api/blob/773578af6334a44c0681e513eca19696ab87b4b1/test/SerialIOTest.cpp#L110), [2](https://github.com/openPMD/openPMD-api/blob/773578af6334a44c0681e513eca19696ab87b4b1/test/SerialIOTest.cpp#L163).

Looking at https://github.com/openPMD/openPMD-api/pull/709 again, I should check again my implementation to see whether there is actually a throwing destructor somewhere (the issue which I fixed in here only came up further down the line in the streaming branch, so we might be fine just letting the destructor run normally). Will look into it tomorrow.

franzpoeschel committed:
Revert "Prevent throwing in destructor in close_iteration_test"

This reverts commit 1490db6577b8ef6f6e517fd1c66773c04f8ad5e0.

franzpoeschel commented:
I don't know exactly where you saw the memory leak – is it gone now? I reverted the leak in the tests now.

franzpoeschel committed:
Add some documentation

ax3l commented on include/openPMD/Iteration.hpp:
```diff
@@ -87,6 +87,43 @@ class Iteration : public Attributable
+    closed() const;
+
+    /**
+     * @brief Has the iteration been closed by the writer?
```
from the description alone I do not understand the difference between `closed()` and `closedByWriter()`.

Do you need to add an additional sentence that states that the *reader* will probe this in streaming context?

ax3l commented on include/openPMD/Iteration.hpp:
```diff
@@ -87,6 +87,43 @@ class Iteration : public Attributable
+     *        Background: Upon calling Iteration::close(), the openPMD API
+     *        will add metadata to the iteration in form of an attribute,
+     *        indicating that the iteration has indeed been closed.
+     *        Useful mainly in streaming context.
```
from the description alone I do not understand the difference between `closed()` and `closedByWriter()`.

Do you need to add an additional sentence that states that the *reader* will probe this in a streaming context?
```suggestion
     *        Useful mainly in streaming context for readers.
```

ax3l replied:
from the description alone I do not understand the difference between `closed()` and `closedByWriter()`.

Do you need to add an additional sentence that states that the *reader* will probe this in a streaming context?
```suggestion
     *        Useful mainly in streaming context when a reader
     *        inquires from a writer that it is done writing.
```

ax3l commented on include/openPMD/Iteration.hpp:
```diff
@@ -99,8 +136,33 @@ class Iteration : public Attributable
     void flush();
     void read();
 
+    enum class CloseStatus
```
Please see https://github.com/openPMD/openPMD-api/blob/beea53937ec1b4774ab087ae24f529d8df0c4cc1/include/openPMD/IO/Access.hpp#L28-L35 for an enum class doxygen string.

ax3l commented on include/openPMD/Iteration.hpp:
```diff
@@ -99,8 +136,33 @@ class Iteration : public Attributable
+     * @return false Otherwise.
+     */
+    bool
+    verifyClosed() const;
```
I would rename this to `closeViolated()` or `stillClosed()` or similar, `verifyClosed()` is not clear. What does this return if an iteration was never closed? The name would also depend on this.

ax3l commented on include/openPMD/Mesh.hpp:
```diff
@@ -182,6 +182,14 @@ class Mesh : public BaseRecord< MeshRecordComponent >
     void flush_impl(std::string const&) override;
     void read() override;
+
+    /**
```
Writing a doxygen string without an element that captures it will propagate the block to the next location where it can "dock". This will lead to undesirable results.
```suggestion
    /*
```

ax3l commented on include/openPMD/Mesh.hpp:
```diff
@@ -182,6 +182,14 @@ class Mesh : public BaseRecord< MeshRecordComponent >
+     *
+     * @return true If closed iteration had no wrong accesses.
+     * @return false Otherwise.
+     */
```
```suggestion
     * @todo needs to be implemented
     */
```

ax3l replied:
is it intentional that particles already have the method but meshes, don't? I think this is missing a declaration here.

franzpoeschel replied:
I implemented the method on the base class instead (i.e. `BaseRecord`) since it also occurs somewhere in the hierarchy for particles. This is dead code, I removed it now.

ax3l commented on include/openPMD/ParticleSpecies.hpp:
```diff
@@ -45,6 +45,16 @@ class ParticleSpecies : public Container< Record >
+     * @return false Otherwise.
+     */
+    bool
+    verifyClosed() const;
```
see comment above about naming and behavior when called on a still-open `ParticleSpecies`

ax3l commented on include/openPMD/RecordComponent.hpp:
```diff
@@ -198,6 +198,16 @@ class RecordComponent : public BaseRecordComponent
+     * @return false Otherwise.
+     */
+    bool
+    verifyClosed() const;
```
see comment above please

ax3l commented on include/openPMD/backend/BaseRecord.hpp:
```diff
@@ -97,7 +97,17 @@ class BaseRecord : public Container< T_elem >
+     * @return false Otherwise.
+     */
+    bool
+    verifyClosed() const;
```
please see above

ax3l commented on src/IO/HDF5/HDF5IOHandler.cpp:
```diff
@@ -399,8 +399,46 @@ HDF5IOHandlerImpl::openFile(Writable* writable,
+    m_openFileIDs.erase( fileID );
+    m_fileIDs.erase( fileID_it );
+
+    /*
+     * std::unordered_map::erase:
```
```suggestion
    /* std::unordered_map::erase:
```

ax3l commented on src/IO/HDF5/HDF5IOHandler.cpp:
```diff
@@ -399,8 +399,46 @@ HDF5IOHandlerImpl::openFile(Writable* writable,
+    auto fileID_it = m_fileIDs.find( writable );
+    if( fileID_it == m_fileIDs.end() )
+    {
+        return;
```
Should that warn? Did you already see that happen? I fear such things will make other problems hard to debug, so just double-checking.

franzpoeschel replied:
I think we can let that throw instead. Closing a file should require that the file is present in the first place.?

ax3l commented on src/Iteration.cpp:
```diff
@@ -19,6 +19,9 @@
  */
 #include "openPMD/Iteration.hpp"
+
+#include <tuple>
```
please include std libs last

ax3l commented on src/Iteration.cpp:
```diff
@@ -97,6 +102,72 @@ Iteration::setTimeUnitSI(double newTimeUnitSI)
+    *m_closed = CloseStatus::ClosedInFrontend;
+    if( _flush )
+    {
+        Series * s = dynamic_cast< Series * >(
```
please validate the `dynamic_cast` is not returning a `nullptr`: #745

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -91,8 +91,95 @@ TEST_CASE( "multi_series_test", "[serial]" )
+    write.flush();
+    // }
+
+    if (isAdios1)
```
There is a new `Series::backend()` aka `AbstractIOHandler::backendName()` that you can query instead.

ax3l commented on src/Iteration.cpp:
```diff
@@ -359,6 +430,30 @@ Iteration::read()
+{
+    if( dirty )
+    {
+        return false;
```
Can you please add three little test cases, e.g. for a single backend, that trigger and assert the `false` logic from an iteration, particle and mesh?

ax3l commented on include/openPMD/Iteration.hpp:
```diff
@@ -99,8 +136,33 @@ class Iteration : public Attributable
+        std::make_shared< CloseStatus >( CloseStatus::Open );
+
+    /**
+     * @brief Verify that a closed iteration has not been wrongly accessed.
```
"wrongly accessed" is too vague, better be concrete what is not allowed:
```suggestion
     * @brief Verify that a closed iteration has not been accessed and/or manipulated further.
```

ax3l commented on include/openPMD/Iteration.hpp:
```diff
@@ -99,8 +136,33 @@ class Iteration : public Attributable
+    /**
+     * @brief Verify that a closed iteration has not been wrongly accessed.
+     *
+     * @return true If closed iteration had no wrong accesses.
```
```suggestion
     * @return true If closed iteration had not further been accessed by the user.
```

franzpoeschel committed:
Add reviewer's suggestions

Also fix a bug in group-based mode

franzpoeschel committed:
Expose Iteration::close to Python API

franzpoeschel committed:
Run simplified test for ADIOS1 in Python

franzpoeschel commented:
I think I added everything from your review now, and additionally Python bindings

ax3l reviewed (approved):
Awesome, thanks a lot! :rocket: :sparkles:
